palace/models/romoperator.cpp

```diff
@@ -294,159 +380,113 @@ void RomOperator::AssemblePROM(double omega)
   {
     RHSr.setZero();
   }
-  if (RHS1.Size())
+  if (has_RHS1)
   {
     RHSr += (1i * omega) * RHS1r;
   }
-}
 
-void RomOperator::SolvePROM(ComplexVector &e)
-{
   // Compute PROM solution at the given frequency and expand into high-dimensional space.
   // The PROM is solved on every process so the matrix-vector product for vector expansion
   // does not require communication.
   BlockTimer bt(Timer::SOLVE_PROM);
-  RHSr = Ar.partialPivLu().solve(RHSr);
-  // RHSr = Ar.ldlt().solve(RHSr);
-  // RHSr = Ar.selfadjointView<Eigen::Lower>().ldlt().solve(RHSr);
-
-  e = 0.0;
-  for (int j = 0; j < dim_V; j++)
-  {
-    e.Real().Add(RHSr(j).real(), V[j]);
-    e.Imag().Add(RHSr(j).imag(), V[j]);
-  }
-}
-
-double RomOperator::ComputeError(double omega)
-{
-  // Compute the error metric associated with the approximate PROM solution at the given
-  // frequency. The HDM residual -r = [K + iω C - ω² M + A2(ω)] x - [iω RHS1 + RHS2(ω)] is
-  // computed using the most recently computed A2(ω) and RHS2(ω).
-  BlockTimer bt(Timer::CONSTRUCT_PROM);
-  AssemblePROM(omega);
-  SolvePROM(w);
-
-  // Residual error.
-  r = 0.0;
-  if (RHS1.Size())
-  {
-    r.Add(-1i * omega, RHS1);
-  }
-  if (has_RHS2)
-  {
-    r.Add(-1.0, RHS2);
-  }
-  double den = !kspKM ? linalg::Norml2(spaceop.GetComm(), r) : 0.0;
-
-  K->AddMult(w, r, 1.0);
-  if (C)
+  if constexpr (false)
   {
-    C->AddMult(w, r, 1i * omega);
-  }
-  M->AddMult(w, r, -omega * omega);
-  if (has_A2)
-  {
-    A2->AddMult(w, r, 1.0);
-  }
-
-  double num;
-  if (!kspKM)
-  {
-    num = linalg::Norml2(spaceop.GetComm(), r);
+    // LDLT solve
+    RHSr = Ar.ldlt().solve(RHSr);
+    RHSr = Ar.selfadjointView<Eigen::Lower>().ldlt().solve(RHSr);
   }
   else
   {
-    z.SetSize(r.Size());
-    z.UseDevice(true);
-    kspKM->Mult(r, z);
-    auto dot = linalg::Dot(spaceop.GetComm(), z, r);
-    MFEM_ASSERT(dot.real() > 0.0 && std::abs(dot.imag()) < 1.0e-9 * dot.real(),
-                "Non-positive vector norm in normalization (dot = " << dot << ")!");
-    num = std::sqrt(dot.real());
-    den = linalg::Norml2(spaceop.GetComm(), w, *kspKM->GetOperator().Real(), z);
+    // LU solve
+    RHSr = Ar.partialPivLu().solve(RHSr);
   }
-  MFEM_VERIFY(den > 0.0, "Unexpected zero denominator in HDM residual!");
-  return num / den;
+  ProlongatePROMSolution(dim_V, V, RHSr, u);
 }
 
-double RomOperator::ComputeMaxError(int num_cand, double &omega_star)
+std::vector<double> RomOperator::FindMaxError(int N) const
 {
```

Comment:
In batch implementations of adaptive sampling, you will want to test (and add) the N worst-case error samples at each iteration. We currently use N = 1, yes, but this function might as well handle the general N case in case someone wants to use a batch method in the future.

Regarding the question, I'm not fully sure I understand but at least yes `z_star` is storing the locations of the N minima of `|Q|` (which, via the reference above, approximate the `N` points of largest PROM error in the parameter domain).

Reply:
I was meaning the non-first minima, i.e. not `[0]` on the output. If the idea is to allow possible batch implementation in future, then that seems reasonable.

Reply:
Oh I see. Yeah I think you always want the global minimum, and it just so happens that you might want also the local ones which follow the first one.